### PR TITLE
Add GL90 parameterization for stacked shallow water

### PR DESCRIPTION
**Pull request by NoraLoose**
### GL90 parameterization

This adds a new vertical viscosity parameterization as in Greatbatch and Lamb (1990), Ferreira & Marshall (2006) and Zhao & Vallis (2008), hereafter referred to as the GL90 vertical viscosity parameterization. This vertical viscosity scheme redistributes momentum in the vertical, and is the equivalent of the Gent & McWilliams (1990) parameterization, but in a TWA (thickness-weighted averaged) set of equations. The vertical viscosity coefficient `nu` is computed from `KD_GL90` (corresponding to `kappa_GM`) via thermal wind balance, and the following relation:

```
nu = KD_GL90 * f^2 / N^2.
```
Currently, the GL90 parameterization is only implemented in stacked shallow water (SSW) mode, in which case we have `1/N^2 = h/g'`.

The vertical viscosity `del_z ( nu del_z u)` is applied to the momentum equation with stress-free boundary conditions at the top and bottom.
    
### GL90 viscosity coefficient: Current options

In the current implementation, `KD_GL90` is assumed either (a) constant or (b) horizontally varying. In both cases, (a) and  (b), one can additionally impose an EBT structure in the vertical for `KD_GL90`. Another possible formulation of nu is depth-independent:

```    
nu = f^2 * alpha_GL90.
```
The latter formulation would be equivalent to a `kappa_GM` that varies as `N^2` with depth.

**Figure**: Vertical viscosity coefficient `Kv_v` for 
**(upper left)** baseline with GL90 parameterization turned off; 
**(upper right)** GL90 parameterization turned on and spatially constant `KD_GL90`; 
**(lower left)** GL90 parameterization turned on and `KD_GL90` which is horizontally constant but varies in the vertical with and EBT structure;
**(lower right)** GL90 parameterization turned on with vertically constant GL90 viscosity, achieved by setting `alpha_GL90 = 3 * 10^7 m^2 s`. 

Note that in the last panel, the diagnostic `Kv_v` is not exactly constant in the vertical because 
* it contains non-GL90 contributions near top and bottom (cf. panel 1)
* it is computed as 
```
f^2 * alpha_GL * ((1/ (h_v(i,k) + h_v(i,k-1)) +  (1/ (h_v(i,k) + h_v(i,k+1)),
```
where `h_v` is the layer thickness used at a v-velocity points in the vertical viscosity scheme. The third factor is the composition of a harmonic and arithmetic mean, and is not equal to the identity.

### Implementation done via coupling coefficients

More specifically, this commit adds a new subroutine that computes the coupling coefficient associated with GL90 via
 
```   
a_cpl_gl90 = nu / h = KD_GL90 * f^2 / g'
```    
 or
 ```  
 a_cpl_gl90 = nu / h = f^2 * alpha_GL90 / h.
```    
Further, `a_cpl_gl90` is multiplied by a function (botfn), which is 0 within the GL90 bottom boundary layer, whose depth is set by `Hbbl_gl90`, and 1 otherwise. This modification is necessary to avoid fluxing momentum into vanished layers that ride over steep topography.
    
Finally, `a_cpl_gl90` is added to `a_cpl`, where the latter is the coupling coefficient associated with the remaining vertical stresses, used in the vertical viscosity solver.
    
More information can be found in Loose et al. (https://www.essoar.org/doi/abs/10.1002/essoar.10512867.1), Appendix B.

### Follow-up PR

There will be a follow-up PR which implements diagnostics related to the GL90 parameterization.

### CODE REVIEW

**Comment by marshallward** [comment]
The log seems to not understand the `ALLOCABLE_` macros:
```
/home/runner/work/MOM6/MOM6/src/parameterizations/vertical/MOM_vert_friction.F90:90: warning: Member allocable_ (variable) of class mom_vert_friction::vertvisc_cs is not documented.
/home/runner/work/MOM6/MOM6/src/parameterizations/vertical/MOM_vert_friction.F90:90: warning: Member dimension (variable) of class mom_vert_friction::vertvisc_cs is not documented.
/home/runner/work/MOM6/MOM6/src/parameterizations/vertical/MOM_vert_friction.F90:90: warning: Member allocable_ (variable) of class mom_vert_friction::vertvisc_cs is not documented.
/home/runner/work/MOM6/MOM6/src/parameterizations/vertical/MOM_vert_friction.F90:90: warning: Member dimension (variable) of class mom_vert_friction::vertvisc_cs is not documented.
```
I'm not sure why it's getting confused here, where there are other working cases, but I will try to have a look soon.  (I'm out of action today and possibly a bit longer, but will try to get to it.)

**Comment by marshallward** [comment]
Oh, I remember a similar issue happening before.  Sometimes a docstring of a previous variable will break, and the error won't manifest until a later line.  Have a look at one of the previous lines.

**Comment by marshallward** [comment]
It's something about the apostrophe `'` in this expression: `1/N^2 = h/g'`

No idea why, but it doesn't like that apostrophe.

**Comment by NoraLoose** [comment]
Nice catch, that fixed it! Thanks. 

It looks like there is a separate issue with codecov.

**Comment by marshallward** [comment]
The codecov one is unfortunately server-side (current theory is they limit the number of uploads within a certain timeframe).  Resubmitting it usually works, though sometimes you have to wait a bit.

**Comment by NoraLoose** [comment]
Ok, we can re-run the github action in a bit?

In any case, I leave this PR in draft mode until #263 is reviewed and merged because this PR depends on it.

**Comment by NoraLoose** [comment]
This PR is ready for review now. @Hallberg-NOAA or @adcroft would be great reviewers as they have seen this code before, while it was being developed.

File: src/parameterizations/vertical/MOM_vert_friction.F90
Comment (Hallberg-NOAA) [inline src/parameterizations/vertical/MOM_vert_friction.F90:1076]: Whenever we add 3 real numbers, we try to add parentheses so that different compilers and compiler settings will give the same answers.  Please add parentheses to specify the order with which the 3 values are being added both here and later on line 1267.

File: src/parameterizations/vertical/MOM_vert_friction.F90
Comment (Hallberg-NOAA) [inline src/parameterizations/vertical/MOM_vert_friction.F90:2115]: Please add arguments like `do_not_log=.not.CSuse_GL90_in_SSW`  to this and any of the following get_param calls that are for any variables that are only used when GL90 is being used, so as to avoid creating overly large MOM_parameter_doc files for cases that are not using GL90 and for which any of these parameters are therefore unused.

**Review by Hallberg-NOAA — CHANGES_REQUESTED** [review]
Overall this excellent PR looks like it is close to being ready to go, but before we merge it into dev/gfdl, I have a few minor specific comments (see below), that I think should be considered.  Thank you for this contribution, @NoraLoose!

File: src/parameterizations/vertical/MOM_vert_friction.F90
Comment (NoraLoose) [reply src/parameterizations/vertical/MOM_vert_friction.F90]: Makes sense - now fixed!

File: src/parameterizations/vertical/MOM_vert_friction.F90
Comment (NoraLoose) [reply src/parameterizations/vertical/MOM_vert_friction.F90]: Good point. I opted to insert an if-statement to not even read all these parameters if `.not. CS%use_GL90_in_SSW`. Saves maybe a few milliseconds? Please let me know if the `do_not_log` approach if preferable for some reason.

Comment (Hallberg-NOAA) [reply src/parameterizations/vertical/MOM_vert_friction.F90]: The overall if statement is fine, and as you note it will be slightly faster.  The only reason why we might prefer the do_not_log approach is that we have the ability to detect unread parameters and optionally bring down the model if there are any unread parameters in the MOM_input or MOM_override files.  Reading these parameters with the do_not_log flag allows us to use an override to the setting of CS%use_GL90_in_SSW to disable the scheme (e.g., in a MOM_override file) without having to comment out any of the subsidiary parameters.  The use of this do_not_log construct is far from universal in the MOM6 code, though, so we are not close to being able to make the default for `FATAL_UNUSED_PARAMS = True`, as we probably should.

I am happy to go with the code as revised, but just be aware that it is possible that at some point we will probably revise this new code as a part of sweeping changes to allow us to change the default for `FATAL_UNUSED_PARAMS`.

Comment (NoraLoose) [reply src/parameterizations/vertical/MOM_vert_friction.F90:2117]: Thanks for the explanation! I switched to the `do_not_log` approach.

Comment (NoraLoose) [reply src/parameterizations/vertical/MOM_vert_friction.F90:2117]: I should mention that for three GL90-related variables I chose `do_not_log = .true.`, rather than `do_not_log = .not. CS%use_GL90_in_SSW`:
* INPUTDIR
* KD_GL90_FILE
* KD_GL90_VARIABLE

I have seen it elsewhere in the code that `INPUTDIR` is not being logged. Not logging the other 2 variables either seemed consistent with that choice. I'm happy to use `do_not_log = .not. CS%use_GL90_in_SSW` for these 3 variables, too, though.

Comment (Hallberg-NOAA) [reply src/parameterizations/vertical/MOM_vert_friction.F90:2117]: INPUTDIR is not being logged in some places because we know that it will be logged elsewhere, and we were trying to manage where it shows up in the MOM_parameter_doc files and does not jump around too much as various module are turned on or off. By contrast, I do not think that the parameters KD_GL90_FILE or KD_GL90_VARIABLE occur anywhere else in the code, so it is important that they are logged here if they are active.

Comment (NoraLoose) [reply src/parameterizations/vertical/MOM_vert_friction.F90:2117]: Gotcha, thanks for clarifying. Changed.

**Review by Hallberg-NOAA — APPROVED** [review]
I am now satisfied that this PR is correct, and I am approving it conditional upon it passing the pipeline testing.

**Comment by Hallberg-NOAA** [comment]
Because of all of the corrections and the merging onto this branch, I am recommending that when this is accepted, it should be done via a squash merge.  @NoraLoose would you object to this?

**Comment by Hallberg-NOAA** [comment]
This PR is undergoing pipeline testing at https://gitlab.gfdl.noaa.gov/ogrp/MOM6/-/pipelines/17774.

**Comment by NoraLoose** [comment]
> Because of all of the corrections and the merging onto this branch, I am recommending that when this is accepted, it should be done via a squash merge. @NoraLoose would you object to this?
@Hallberg-NOAA sounds good.

**Comment by marshallward** [comment]
Gaea regression: https://gitlab.gfdl.noaa.gov/ogrp/MOM6/-/pipelines/17774 :heavy_check_mark: :yellow_circle:

**Pull request merged**
